### PR TITLE
Create an Analysis Request from Sample View

### PR DESCRIPTION
**Pull request by Espurna**
## Description of the issue/feature this PR addresses

This PR adds a button in Sample Views to create analysis requests from them.

An “Add Analysis Request” button will be added in Sample's view. Once the button is pressed, the user will be redirected to the Analysis Request Add form, and the data already assigned to the Sample will be filled automatically

Since some AnalysisRequest object fields can be overridden or extended in addons, a marker interface and an adapters query has been added to `get_default_value`. The process is as follows:

Default data for AR Add form fields are filed through `get_default_value` method that give them a value using some logic. In some cases those fields can be defined in addons.  Now `get_default_value` queries for hooks adapting `IGetDefaultFieldValueARAddHook`, those hooks should return a default value for the extended fields.

## Current behavior before PR

The only way to create an analysis request is from an analysis request view. If a user is checking a sample and wants to create an analysis request for it, he/she has to move to analysis requests view, open a new AR Add form and introduce the sample ID in the AR Add form field.

## Desired behavior after PR is merged

An analysis request for a sample can be created from the same Sample View, easing the process.

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/browser/analysisrequest/add2.py [hunk 1]
```diff
@@ -320,6 +324,15 @@ def get_client(self):
             return context.getClient()
         return None
 
+    def get_sample(self):
+        """
+        Returns the Sample
+        """
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/add2.py]: I know, it might sound pedantic, but stick to the same docstring style as the other methods, e.g.
```python
def get_sample(self):
    """Returns the Sample
    """
```

File: bika/lims/browser/sample/edit.py [hunk 2]
```diff
@@ -7,18 +7,19 @@
...
 from bika.lims.browser.header_table import HeaderTableView
 from bika.lims.config import POINTS_OF_CAPTURE
-from bika.lims.permissions import *
+from bika.lims.permissions import AddAnalysisRequest
 from bika.lims.workflow import doActionFor
-from plone.app.layout.globals.interfaces import IViewView
-from zope.interface import implements
-from . import SamplePartitionsView
 from . import SampleAnalysesView
+from . import SamplePartitionsView
```
Comment (ramonski) [inline bika/lims/browser/sample/edit.py]: Try to put PEP8 changes into a separate commit

File: bika/lims/browser/sample/templates/sample.pt [hunk 3]
```diff
@@ -9,10 +9,18 @@
...
+		<tal:add_actions repeat="add_item python:view.context_actions.keys()">
+			<a tal:attributes="
+				style python:'background: url(%s) 2px 50%% no-repeat'%(view.context_actions[add_item]['icon']);
+			 	href python:view.context_actions[add_item]['url'];
+			 	class python:'context_action_link %s' % (view.context_actions[add_item].get('class',''))">
+				<span tal:replace="python:add_item"/>
+			</a>
+		</tal:add_actions>
+	</h1>
 </metal:title>
```
Comment (ramonski) [inline bika/lims/browser/sample/templates/sample.pt]: Try to format also HTML files nicely, e.g. with an indendation of 2 spaces to safe some space. Also avoid whenever possible inline styles

File: bika/lims/browser/analysisrequest/add2.py [hunk 4]
```diff
@@ -6,30 +6,28 @@
...
+from zope.interface import implements
+from zope.publisher.interfaces import IPublishTraverse
 
 from bika.lims import api
-from bika.lims import logger
 from bika.lims import bikaMessageFactory as _
+from bika.lims import logger
+from bika.lims.interfaces import IGetDefaultFieldValueARAddHook
 from bika.lims.utils import tmpID
 from bika.lims.utils.analysisrequest import create_analysisrequest as crar
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/add2.py]: Here also a separate commit please

File: bika/lims/browser/analysisrequest/add2.py [hunk 5]
```diff
@@ -296,6 +294,24 @@ def get_default_value(self, field, context):
...
+        #   provides = "bika.lims.interfaces.IGetDefaultFieldValueARAddHook"
+        #   name = "<fieldName>_default_value_hook"
+        # / >
+        hook_name = name + '_default_value_hook'
+        adapter = queryAdapter(
+            self.request,
+            name=hook_name,
+            interface=IGetDefaultFieldValueARAddHook)
+        if adapter is not None:
+            default = adapter(self.context)
```
Comment (ramonski) [inline bika/lims/browser/analysisrequest/add2.py]: Nice!

File: bika/lims/browser/js/coffee/bika.lims.analysisrequest.add.coffee [hunk 6]
```diff
@@ -541,11 +541,16 @@ class window.AnalysisRequestAdd
...
+    # set the sampling date
     field = $("#SamplingDate-#{arnum}")
     value = sample.sampling_date
     field.val value
 
+    # set the date sampeld
+    field = $("#DateSampled-#{arnum}")
+    value = sample.date_sampled
+    field.val value
+
```
Comment (ramonski) [inline bika/lims/browser/js/coffee/bika.lims.analysisrequest.add.coffee]: Try to keep your commits consistent. E.g. this change is independent of the hook you introduced above

**Review by ramonski — CHANGES_REQUESTED** [review]
Thanks Pau, I like the idea with the adapter to fetch the default value – nicely decoupled.
I just have some change request according to formatting and PEP8-ing, sorry for being nerving with that, but these small things are important for my little coders heart ;)

File: bika/lims/browser/analysisrequest/add2.py (hunk 4 above)
Comment (Espurna) [reply bika/lims/browser/analysisrequest/add2.py]: What do you mean? A separate commit for import modifications?

File: bika/lims/browser/js/coffee/bika.lims.analysisrequest.add.coffee (hunk 6 above)
Comment (Espurna) [reply bika/lims/browser/js/coffee/bika.lims.analysisrequest.add.coffee]: Got it

File: bika/lims/browser/sample/templates/sample.pt (hunk 3 above)
Comment (Espurna) [reply bika/lims/browser/sample/templates/sample.pt]: Done, pycharm was using tabs as default

File: bika/lims/browser/analysisrequest/add2.py (hunk 4 above)
Comment (Espurna) [reply bika/lims/browser/analysisrequest/add2.py]: Ok

File: bika/lims/browser/sample/edit.py (hunk 2 above)
Comment (Espurna) [reply bika/lims/browser/sample/edit.py]: Done!

**Review by xispa — APPROVED** [review]
Added entry to CHANGES.rst, removed duplicate code and merged master into
